engine/src/execution_graph/logic_controllers/taskflow/graph.cpp

```diff
@@ -0,0 +1,255 @@
+#include "graph.h"
+
+namespace ral {
+namespace cache {
+
+	kpair graph::operator+=(kpair p) {
+		std::string source_port_name = std::to_string(p.src->get_id());
+		std::string target_port_name = std::to_string(p.dst->get_id());
+
+		if(p.has_custom_source()) {
+			source_port_name = p.src_port_name;
+		}
+		if(p.has_custom_target()) {
+			target_port_name = p.dst_port_name;
+		}
+		this->add_edge(p.src, p.dst, source_port_name, target_port_name, p.cache_machine_config);
+		return p;
+	}
+
+	void graph::check_and_complete_work_flow() {
```

Comment:
When is this function used?

Reply:
This is called at the beginning of graph::execute(). It basically cleans up the graph, by tying up so to speak the ends where there is like a dangling edge.
This function already existed and i did not touch it. Maybe there was a whitespace change

Reply:
What do you mean see if there is a dangling edge? Do you mean see if the graph was made improperly? If it was made improperly how does it recover?

Reply:
No its more about the first or last node (i cant remember which). They need special handling. THat is what this function does. Sorry my original explanation was not very good